libfwupd/fwupd-client.c

```diff
@@ -4228,56 +4231,131 @@ fwupd_client_download_write_callback_cb (char *ptr, size_t size, size_t nmemb, v
+	g_debug ("downloading %s", url);
+	fwupd_client_set_status (self, FWUPD_STATUS_DOWNLOADING);
+	curl_easy_setopt (curl, CURLOPT_URL, url);
+	curl_easy_setopt (curl, CURLOPT_ERRORBUFFER, errbuf);
+	curl_easy_setopt (curl, CURLOPT_WRITEFUNCTION, fwupd_client_download_write_callback_cb);
+	curl_easy_setopt (curl, CURLOPT_WRITEDATA, buf);
+	res = curl_easy_perform (curl);
 	fwupd_client_set_status (self, FWUPD_STATUS_IDLE);
 	if (res != CURLE_OK) {
 		glong status_code = 0;
-		curl_easy_getinfo (helper->curl, CURLINFO_RESPONSE_CODE, &status_code);
+		curl_easy_getinfo (curl, CURLINFO_RESPONSE_CODE, &status_code);
 		g_debug ("status-code was %ld", status_code);
 		if (status_code == 429) {
-			g_task_return_new_error (task,
-						 FWUPD_ERROR,
-						 FWUPD_ERROR_INVALID_FILE,
-						 "Failed to download due to server limit");
-			return;
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "Failed to download due to server limit");
+			return NULL;
 		}
 		if (errbuf[0] != '\0') {
-			g_task_return_new_error (task,
-						 FWUPD_ERROR,
-						 FWUPD_ERROR_INVALID_FILE,
-						 "failed to download file: %s",
-						 errbuf);
-			return;
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "failed to download file: %s",
+				     errbuf);
+			return NULL;
 		}
-		g_task_return_new_error (task,
-					 FWUPD_ERROR,
-					 FWUPD_ERROR_INVALID_FILE,
-					 "failed to download file: %s",
-					 curl_easy_strerror (res));
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INVALID_FILE,
+			     "failed to download file: %s",
+			     curl_easy_strerror (res));
+		return NULL;
+	}
+	return g_byte_array_free_to_bytes (g_steal_pointer (&buf));
+}
 
-		return;
+static void
+fwupd_client_download_bytes_thread_cb (GTask *task,
+				       gpointer source_object,
+				       gpointer task_data,
+				       GCancellable *cancellable)
+{
+	FwupdClient *self = FWUPD_CLIENT (source_object);
+	FwupdCurlHelper *helper = g_task_get_task_data (task);
+	g_autoptr(GBytes) blob = NULL;
+
+	for (guint i = 0; i < helper->urls->len; i++) {
+		const gchar *url = g_ptr_array_index (helper->urls, i);
+		g_autoptr(GError) error = NULL;
+		if (fwupd_client_is_url_http (url)) {
+			blob = fwupd_client_download_http (self, helper->curl, url, &error);
+			if (blob != NULL)
+				break;
+		} else {
+			g_set_error (&error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "not sure how to handle: %s", url);
+		}
+		if (i == helper->urls->len - 1) {
+			g_task_return_error (task, g_steal_pointer (&error));
+			return;
+		}
+		g_warning ("failed to download %s: %s, trying next URI…",
+			   url, error->message);
 	}
 	g_task_return_pointer (task,
-			       g_byte_array_free_to_bytes (g_steal_pointer (&buf)),
+			       g_steal_pointer (&blob),
 			       (GDestroyNotify) g_bytes_unref);
 }
 #endif
 
+/* private */
+void
+fwupd_client_download_bytes2_async (FwupdClient *self,
```

Comment:
presumably this should change with the ABI bump and not be the private one anymore.

Reply:
Yes, probably; let's see if we need any more parameters before we export it.